onnxruntime/test/onnx/runner.cc

```diff
@@ -489,30 +511,30 @@ void RunSingleTestCase(const ITestCase& info, Ort::Env& env, const Ort::SessionO
       concurrent_runs = 1;
     }
 
+    DataRunner* r;
     if (concurrent_runs > 1 && data_count > 1) {
-      r.reset(new PTestRunner(session_object.release(), info, tpool, on_finished));
+      r = new PTestRunner(session_object.release(), info, tpool, on_finished);
     } else {
-      r.reset(new SeqTestRunner(session_object.release(), info, repeat_count, on_finished));
+      r = new SeqTestRunner(session_object.release(), info, repeat_count, on_finished);
     }
     r->Start(pci, concurrent_runs);
 
-    // DataRunner::Finish will delete itself, so now that we know everything has started without any exceptions
-    // we can release it from the unique_ptr
-    r.release();
+    // both PTestRunner and SeqTestRunner call DataRunner::Finish which will delete itself and call on_finished callback
```

Comment:
If Start() throws, will 'r' leak?

Reply:
No, no leaks. Start will not throw. The crash was because of the dual nature of destruciton: 1) via unique_ptr 2) via self callback

---
In reply to: [452674635](https://github.com/microsoft/onnxruntime/pull/4475#discussion_r452674635) [](ancestors = 452674635)

Reply:
Ok, again it's not obvious. Adding a one line comment might help. Can be in a separate PR.

Reply:
I added comments to explain the ownership of DataRunner on line 517